service/history/shard/task_key_manager.go

```diff
@@ -125,6 +125,9 @@ func (m *taskKeyManager) getExclusiveReaderHighWatermark(
 	)
 
 	nextTaskKey := m.generator.peekTaskKey(category)
+	if category.ID() == tasks.CategoryIDReplication {
```

Comment:
we share task_id across different category. This seems like we want to have a special task id serial for replication.

Reply:
For exclusiveReaderHighWatermark, we don't need to use special task id serial for replication. Having separated readerHighWatermark is enough. Eventually this can apply to all categories: each category uses its own readerHighWatermark.

Reply:
Not exactly sure what caused the flakyness. It is because replication ackManager only updates to the last replicated taskID?

If the taskID is allocated but the failed to be persisted, we will run into the same issue right?

Reply:
Non-replication tasks such as timers could happen anytime which leads to the error scenario: the last task is non-replication task.

And yes, replication ackManager only updates to the last replicated taskID ([code](https://github.com/temporalio/temporal/blob/37d8bba34caaed5fb3aff3b25021da18a8290cd6/service/history/history_engine.go#L767)).

"If the taskID is allocated but the failed to be persisted" -- I'm not sure the exact case you mentioned. From the code `taskKeyGenerator` will get a new range after initialization, so new taskID will be larger than any of previous one. This should not be a problem.